cryptographic_estimators/UOVEstimator/UOVAlgorithms/collision_attack.py

```diff
@@ -0,0 +1,170 @@
+    - ``complexity_type`` -- complexity type to consider (0: estimate, 1: tilde O complexity, default: 0)
+
+    """
+
+    def __init__(self, problem: UOVProblem, **kwargs):
+        super().__init__(problem, **kwargs)
+
+        self._name = "CollisionAttack"
+        self._attack_type = "forgery"
+        self._alpha = 1.25
+        q = problem.order_of_the_field()
+        self._gray_code_eval_cost = kwargs.get("gray_code_eval_cost", log2(q))
+
+
+    @optimal_parameter
+    def X(self):
+        """
+        Return the optimal `X`, i.e. no. of inputs (preimages)
+
+        EXAMPLES::
+
+            sage: from cryptographic_estimators.UOVEstimator.UOVAlgorithms.collision_attack import CollisionAttack
+            sage: from cryptographic_estimators.UOVEstimator.uov_problem import UOVProblem
+            sage: E = CollisionAttack(UOVProblem(n=24, m=10, q=2))
+            sage: E.X()
+            11.95
+
+        """
+        _, m, q = self.problem.get_parameters()
+        alpha = self._alpha
+        r = self._gray_code_eval_cost
+        X = 9.25  +  (1 / 2) * (-log2(3 * m * r)  + log2(alpha) + m * log2(q))
+        precision = 2
+        X_truncated = float(int(X * 10 ** precision) / 10 ** precision)
+        return X_truncated
+    
+    @optimal_parameter
+    def Y(self):
+        """
+        Return logarithm of the optimal `Y`, i.e. logarithm of no. of hashes to compute
+
+        EXAMPLES::
+
+            sage: from cryptographic_estimators.UOVEstimator.UOVAlgorithms.collision_attack import CollisionAttack
+            sage: from cryptographic_estimators.UOVEstimator.uov_problem import UOVProblem
+            sage: E = CollisionAttack(UOVProblem(n=24, m=10, q=2))
+            sage: E.Y()
+            0.0
+
+        """
+        X =  self.X()
+        alpha = self._alpha
+        _, m, q = self.problem.get_parameters()
+        Y = max(log2(alpha) + m * log2(q) - X, 0)
+        precision = 2
+        Y_truncated = float(int(Y * 10 ** precision) / 10 ** precision)
+        return Y_truncated
+
+    def _compute_time_complexity(self, parameters: dict):
+        """
+        Return the time complexity of the algorithm for a given set of parameters
+
+        INPUT:
+
+        - ``parameters`` -- dictionary including the parameters
+
+        TESTS::
+
+            sage: from cryptographic_estimators.UOVEstimator.UOVAlgorithms.collision_attack import CollisionAttack
+            sage: from cryptographic_estimators.UOVEstimator.uov_problem import UOVProblem
+            sage: E = CollisionAttack(UOVProblem(n=24, m=10, q=2))
+            sage: E.time_complexity()
+            17.387737896811494
+
+        """
+        _, m, q = self.problem.get_parameters()
+        X = self.X()
+        Y = self.Y()
+        alpha = self._alpha
+        r = self._gray_code_eval_cost
+        time_temp = int((3 / 2) * m * r * 2 ** X) + int(2 ** 17.5 * 2 ** Y)
```

Comment:
is the `int` cast necessary? Why not just work with floats, better logarithmic floats?

Reply:
Maybe not but how would you compute `log2(2**X + 2**Y)` given `X` and `Y`?